File: packages/core/src/device-contract.ts
import { ProtocolError } from "./protocol-error.js";
import {
  PROTOCOL_VERSION,
  createRequestId,
  hasOnlyObjectKeys,
  isRecord,
} from "./protocol-primitives.js";
import { isSafeRequestId, isSessionId } from "./safe-text.js";

export type DeviceSessionRule = "forbidden" | "required" | "optional";
export type DevicePayloadRule = "forbidden" | "required" | "optional";

export interface DeviceMethodRow {
  readonly method: string;
  readonly sessionRule: DeviceSessionRule;
  readonly payloadRule: DevicePayloadRule;
  readonly payloadSchemaOwner: string;
  readonly resultSchemaOwner: string;
  readonly firmwareGate: string;
}

export const DEVICE_METHOD_ROWS = [
  {
    method: "get_status",
    sessionRule: "forbidden",
    payloadRule: "forbidden",
    payloadSchemaOwner: "no payload",
    resultSchemaOwner: "StatusResult",
    firmwareGate: "status_material_consistency",
  },
  {
    method: "identify_device",
    sessionRule: "forbidden",
    payloadRule: "required",
    payloadSchemaOwner: "IdentifyDevicePayload",
    resultSchemaOwner: "IdentifyDeviceResult",
    firmwareGate: "identification_display",
  },
  {
    method: "connect",
    sessionRule: "forbidden",
    payloadRule: "required",
    payloadSchemaOwner: "ConnectPayload",
    resultSchemaOwner: "ConnectResult",
    firmwareGate: "provisioned_material_policy_auth_approval",
  },
  {
    method: "disconnect",
    sessionRule: "required",
    payloadRule: "forbidden",
    payloadSchemaOwner: "no payload",
    resultSchemaOwner: "DisconnectResult",
    firmwareGate: "active_session_cleanup",
  },
  {
    method: "get_capabilities",
    sessionRule: "required",
    payloadRule: "forbidden",
    payloadSchemaOwner: "no payload",
    resultSchemaOwner: "CapabilitiesResult",
    firmwareGate: "active_session_provisioned_material",
  },
  {
    method: "get_accounts",
    sessionRule: "required",
    payloadRule: "forbidden",
    payloadSchemaOwner: "no payload",
    resultSchemaOwner: "AccountsResult",
    firmwareGate: "active_session_account_material",
  },
  {
    method: "policy_get",
    sessionRule: "required",
    payloadRule: "forbidden",
    payloadSchemaOwner: "no payload",
    resultSchemaOwner: "PolicyGetResult",
    firmwareGate: "active_session_policy_store",
  },
  {
    method: "get_approval_history",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "ApprovalHistoryPayload",
    resultSchemaOwner: "ApprovalHistoryResult",
    firmwareGate: "active_session_approval_history_store",
  },
  {
    method: "sign_transaction",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "SuiSignTransactionPayload",
    resultSchemaOwner: "SignTransactionResult",
    firmwareGate: "active_session_account_binding_signing_authorization",
  },
  {
    method: "sign_personal_message",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "SuiSignPersonalMessagePayload",
    resultSchemaOwner: "SignPersonalMessageResult",
    firmwareGate: "active_session_account_binding_user_confirmation",
  },
  {
    method: "policy_propose",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "PolicyProposePayload",
    resultSchemaOwner: "PolicyProposalOutcome",
    firmwareGate: "active_session_policy_review_auth_commit",
  },
  {
    method: "credential_prepare",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "CredentialPreparePayload",
    resultSchemaOwner: "CredentialPreparation",
    firmwareGate: "active_session_credential_preparation",
  },
  {
    method: "credential_propose",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "CredentialProposePayload",
    resultSchemaOwner: "CredentialProposalOutcome",
    firmwareGate: "active_session_credential_review_auth_commit",
  },
  {
    method: "get_result",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "RetainedResponsePayload",
    resultSchemaOwner: "retained method response",
    firmwareGate: "active_session_retained_response_lookup",
  },
  {
    method: "ack_result",
    sessionRule: "required",
    payloadRule: "required",
    payloadSchemaOwner: "RetainedResponsePayload",
    resultSchemaOwner: "AckResult",
    firmwareGate: "active_session_retained_response_cleanup",
  },
] as const satisfies readonly DeviceMethodRow[];

export type DeviceMethod = (typeof DEVICE_METHOD_ROWS)[number]["method"];

const DEVICE_METHOD_SET = new Set<string>(DEVICE_METHOD_ROWS.map((row) => row.method));
const DEVICE_METHOD_ROW_BY_METHOD = new Map<DeviceMethod, (typeof DEVICE_METHOD_ROWS)[number]>(
  DEVICE_METHOD_ROWS.map((row) => [row.method, row]),
);

export function isDeviceMethod(value: unknown): value is DeviceMethod {
  return typeof value === "string" && DEVICE_METHOD_SET.has(value);
}

export function deviceMethodRow(method: DeviceMethod): (typeof DEVICE_METHOD_ROWS)[number] {
  const row = DEVICE_METHOD_ROW_BY_METHOD.get(method);
  if (row === undefined) {
    throw new ProtocolError("unsupported_method", "Device method is not supported.");
  }
  return row;
}

export interface DeviceErrorRow {
  readonly code: string;
  readonly retryable: boolean;
  readonly message: string;
  readonly meaning: string;
}

export const DEVICE_ERROR_ROWS = [
  { code: "invalid_request", retryable: false, message: "Device request is malformed.", meaning: "Envelope, JSON, id, version, or method shape is invalid." },
  { code: "invalid_params", retryable: false, message: "Device request payload is invalid.", meaning: "Method payload shape or value is invalid." },
  { code: "invalid_state", retryable: false, message: "Device state does not allow this request.", meaning: "Firmware state does not allow the method." },
  { code: "invalid_session", retryable: false, message: "Session is missing, expired, or does not match.", meaning: "Session is missing, expired, or does not match." },
  { code: "request_id_conflict", retryable: false, message: "Request id is already bound to a different request.", meaning: "Request id is already bound to a different retained request identity." },
  { code: "unknown_request", retryable: false, message: "Requested retained response does not exist.", meaning: "Requested retained response does not exist in the current session." },
  { code: "no_active_device", retryable: false, message: "No active device is configured.", meaning: "Host process has no selected active device for the requested scope." },
  { code: "device_not_found", retryable: true, message: "Requested device is not known to Agent-Q.", meaning: "Host process cannot find the requested stored device." },
  { code: "invalid_device_id", retryable: false, message: "Device id is invalid.", meaning: "Host-side device id input is invalid." },
  { code: "invalid_device", retryable: false, message: "Device identity is unsafe.", meaning: "Host process rejected a device identity as unsafe." },
  { code: "device_mismatch", retryable: false, message: "Connected device does not match the requested device.", meaning: "Host process connected to a device whose Firmware identity did not match the requested device id." },
  { code: "unsupported_method", retryable: false, message: "Device method is not supported.", meaning: "Method is not implemented by this Firmware or host boundary." },
  { code: "unsupported_version", retryable: false, message: "Protocol version is not supported.", meaning: "Protocol version is not supported." },
  { code: "unsupported_chain", retryable: false, message: "Chain is not supported.", meaning: "Chain is not supported for the method." },
  { code: "unsupported_transaction", retryable: false, message: "Transaction shape is not supported.", meaning: "Transaction shape is not supported by the current signing path." },
  { code: "malformed_transaction", retryable: false, message: "Transaction bytes are malformed.", meaning: "Transaction bytes or transaction structure are malformed." },
  { code: "payload_too_large", retryable: false, message: "Payload exceeds the current device payload capacity.", meaning: "Payload exceeds the current device payload capacity." },
  { code: "payload_unavailable", retryable: false, message: "Referenced payload is unavailable.", meaning: "Referenced internal payload is missing, expired, consumed, or wrong-session." },
  { code: "payload_conflict", retryable: true, message: "Another sensitive request blocks this payload.", meaning: "Another payload or sensitive flow blocks this request." },
  { code: "busy", retryable: true, message: "Device is busy with another request.", meaning: "Device is already handling another request." },
  { code: "timeout", retryable: true, message: "The signing request timed out on the device.", meaning: "Device or user action timed out." },
  { code: "user_rejected", retryable: false, message: "The signing request was rejected on the device.", meaning: "Device-local user confirmation rejected the request." },
  { code: "policy_rejected", retryable: false, message: "The signing request was rejected by device policy.", meaning: "Firmware policy rejected the request." },
  { code: "signing_failed", retryable: false, message: "The device could not produce a signature.", meaning: "Firmware could not produce the requested signature." },
  { code: "ui_error", retryable: false, message: "Device could not display or manage required UI.", meaning: "Firmware could not display or manage required device UI." },
  { code: "auth_unavailable", retryable: false, message: "Device-local authentication verifier is unavailable.", meaning: "Device-local authentication verifier is unavailable." },
  { code: "account_unavailable", retryable: false, message: "Device account material is unavailable or does not match the request.", meaning: "Device account material is unavailable or does not match the request." },
  { code: "policy_unavailable", retryable: false, message: "Firmware policy store or active policy is unavailable.", meaning: "Firmware policy store or active policy is unavailable." },
  { code: "history_unavailable", retryable: false, message: "Approval history is unavailable.", meaning: "Approval history is unavailable." },
  { code: "rng_unavailable", retryable: true, message: "Device random generator is unavailable.", meaning: "Firmware random generator is unavailable." },
  { code: "invalid_response", retryable: true, message: "Device response is malformed.", meaning: "Firmware response JSON, envelope, or method result shape is invalid." },
  { code: "handshake_failed", retryable: true, message: "Device status or identification handshake failed.", meaning: "Device status or identification handshake failed." },
  { code: "port_not_found", retryable: true, message: "Requested device port is not connected.", meaning: "Requested device port is not connected." },
  { code: "port_in_use", retryable: true, message: "Requested device port is already in use.", meaning: "Requested device port is already in use." },
  { code: "port_permission_denied", retryable: false, message: "Host process lacks permission to access the device port.", meaning: "Host process lacks permission to access the device port." },
  { code: "unsupported_transport", retryable: false, message: "Host environment does not support the required device transport.", meaning: "Host environment does not provide the transport required by the adapter." },
  { code: "transport_closed", retryable: true, message: "Device connection closed before a valid response.", meaning: "Established host-device connection closed before a valid response." },
  { code: "transport_error", retryable: true, message: "Host-device transport failed before a valid response.", meaning: "Host-device transport failed before a valid device response was received." },
  { code: "local_server_unavailable", retryable: true, message: "Local Agent-Q server is unavailable.", meaning: "Local CLI client cannot reach the local Agent-Q server." },
  { code: "identification_code_exhausted", retryable: true, message: "Identification code could not be allocated.", meaning: "Host process could not allocate a unique identification code." },
  { code: "internal_output_error", retryable: false, message: "Agent-Q produced an unexpected internal result.", meaning: "Output sanitization or internal result validation failed." },
  { code: "unknown_error", retryable: false, message: "Agent-Q request failed.", meaning: "Failure did not match a known public code." },
] as const satisfies readonly DeviceErrorRow[];

export type DeviceErrorCode = (typeof DEVICE_ERROR_ROWS)[number]["code"];

const DEVICE_ERROR_ROW_BY_CODE = new Map<DeviceErrorCode, (typeof DEVICE_ERROR_ROWS)[number]>(
  DEVICE_ERROR_ROWS.map((row) => [row.code, row]),
);
const DEVICE_ERROR_CODE_SET = new Set<string>(DEVICE_ERROR_ROWS.map((row) => row.code));

export interface DeviceError {
  readonly code: DeviceErrorCode;
  readonly message: string;
  readonly retryable: boolean;
}

export function isDeviceErrorCode(value: unknown): value is DeviceErrorCode {
  return typeof value === "string" && DEVICE_ERROR_CODE_SET.has(value);
}

export function deviceErrorRow(code: DeviceErrorCode): (typeof DEVICE_ERROR_ROWS)[number] {
  const row = DEVICE_ERROR_ROW_BY_CODE.get(code);
  if (row === undefined) {
    throw new ProtocolError("invalid_response", "Device error code is not in the contract table.");
  }
  return row;
}

export function makeDeviceError(code: DeviceErrorCode): DeviceError {
  const row = deviceErrorRow(code);
  return { code: row.code, message: row.message, retryable: row.retryable };
}

export function makeUnknownDeviceError(): DeviceError {
  return makeDeviceError("unknown_error");
}

const SIGNING_OUTCOME_STATUS_DEVICE_ERROR_CODES = {
  user_rejected: "user_rejected",
  user_timed_out: "timeout",
  policy_rejected: "policy_rejected",
  signing_failed: "signing_failed",
} as const satisfies Record<string, DeviceErrorCode>;

export type SigningOutcomeErrorCode = keyof typeof SIGNING_OUTCOME_STATUS_DEVICE_ERROR_CODES;

function signingOutcomeErrorMessage(code: SigningOutcomeErrorCode): string {
  return deviceErrorRow(SIGNING_OUTCOME_STATUS_DEVICE_ERROR_CODES[code]).message;
}

export const SIGNING_OUTCOME_ERROR_MESSAGES = {
  user_rejected: signingOutcomeErrorMessage("user_rejected"),
  user_timed_out: signingOutcomeErrorMessage("user_timed_out"),
  policy_rejected: signingOutcomeErrorMessage("policy_rejected"),
  signing_failed: signingOutcomeErrorMessage("signing_failed"),
} as const satisfies Record<SigningOutcomeErrorCode, string>;

export interface DeviceRequest {
  readonly id: string;
  readonly version: typeof PROTOCOL_VERSION;
  readonly sessionId?: string;
  readonly method: DeviceMethod;
  readonly payload?: unknown;
}

export interface MakeDeviceRequestInput {
  readonly id?: string;
  readonly sessionId?: string;
  readonly method: DeviceMethod;
  readonly payload?: unknown;
}

export function makeDeviceRequest(input: MakeDeviceRequestInput): DeviceRequest {
  const request: Record<string, unknown> = {
    id: input.id ?? createRequestId(),
    version: PROTOCOL_VERSION,
    method: input.method,
  };
  if (input.sessionId !== undefined) {
    request.sessionId = input.sessionId;
  }
  if (Object.prototype.hasOwnProperty.call(input, "payload")) {
    request.payload = input.payload;
  }
  return parseDeviceRequest(request);
}

export function parseDeviceRequest(value: unknown): DeviceRequest {
  const record = asContractRecord(value, "DeviceRequest", "invalid_request");
  if (!hasOnlyObjectKeys(record, ["id", "version", "sessionId", "method", "payload"])) {
    throw new ProtocolError("invalid_request", "Device request contains unsupported fields.");
  }
  if (!isSafeRequestId(record.id)) {
    throw new ProtocolError("invalid_request", "Device request id is invalid.");
  }
  if (typeof record.version !== "number" || !Number.isInteger(record.version)) {
    throw new ProtocolError("invalid_request", "Device request version is invalid.");
  }
  if (record.version !== PROTOCOL_VERSION) {
    throw new ProtocolError("unsupported_version", "Protocol version is not supported.");
  }
  if (typeof record.method !== "string") {
    throw new ProtocolError("invalid_request", "Device request method is invalid.");
  }
  if (!isDeviceMethod(record.method)) {
    throw new ProtocolError("unsupported_method", "Device method is not supported.");
  }
  const row = deviceMethodRow(record.method);
  const hasSessionId = Object.prototype.hasOwnProperty.call(record, "sessionId");
  const hasPayload = Object.prototype.hasOwnProperty.call(record, "payload");
  if (hasPayload && record.payload === undefined) {
    throw new ProtocolError("invalid_params", "Device request payload must not be undefined.");
  }
  validateSessionRule(row.sessionRule, hasSessionId, record.sessionId);
  validatePayloadRule(row.payloadRule, hasPayload);

  return {
    id: record.id,
    version: PROTOCOL_VERSION,
    ...(hasSessionId ? { sessionId: record.sessionId as string } : {}),
    method: record.method,
    ...(hasPayload ? { payload: record.payload } : {}),
  };
}

export function serializeDeviceRequest(request: DeviceRequest): string {
  return `${JSON.stringify(parseDeviceRequest(request))}\n`;
}

export type DeviceResultParser<T = unknown> = (result: unknown) => T;

export interface DeviceSuccessResponse<TResult = unknown> {
  readonly id?: string;
  readonly version: typeof PROTOCOL_VERSION;
  readonly success: true;
  readonly method: DeviceMethod;
  readonly result: TResult;
}

export interface DeviceFailureResponse {
  readonly id?: string;
  readonly version: typeof PROTOCOL_VERSION;
  readonly success: false;
  readonly method?: DeviceMethod;
  readonly error: DeviceError;
}

export type DeviceResponse<TResult = unknown> = DeviceSuccessResponse<TResult> | DeviceFailureResponse;

export interface ParseDeviceResponseOptions {
  readonly expectedId?: string;
  readonly expectedMethod?: DeviceMethod;
  readonly resultParsers?: Partial<Record<DeviceMethod, DeviceResultParser>>;
}

export function makeDeviceSuccessResponse<TResult>(
  input: Omit<DeviceSuccessResponse<TResult>, "version" | "success">,
): DeviceSuccessResponse<TResult> {
  return parseDeviceResponse({ ...input, version: PROTOCOL_VERSION, success: true }) as DeviceSuccessResponse<TResult>;
}

export function makeDeviceFailureResponse(input: {
  readonly id?: string;
  readonly method?: DeviceMethod;
  readonly code: DeviceErrorCode;
}): DeviceFailureResponse {
  const response: Record<string, unknown> = {
    version: PROTOCOL_VERSION,
    success: false,
    error: makeDeviceError(input.code),
  };
  if (input.id !== undefined) {
    response.id = input.id;
  }
  if (input.method !== undefined) {
    response.method = input.method;
  }
  return parseDeviceResponse(response) as DeviceFailureResponse;
}

export function parseDeviceResponse<TResult = unknown>(
  value: unknown,
  options: ParseDeviceResponseOptions = {},
): DeviceResponse<TResult> {
  const record = asContractRecord(value, "DeviceResponse", "invalid_response");
  if (!hasOnlyObjectKeys(record, ["id", "version", "success", "method", "result", "error"])) {
    throw new ProtocolError("invalid_response", "Device response contains unsupported fields.");
  }
  if (Object.prototype.hasOwnProperty.call(record, "id") && !isSafeRequestId(record.id)) {
    throw new ProtocolError("invalid_response", "Device response id is invalid.");
  }
  if (options.expectedId !== undefined && record.id !== options.expectedId) {
    throw new ProtocolError("invalid_response", "Device response id does not match request.");
  }
  if (record.version !== PROTOCOL_VERSION) {
    throw new ProtocolError("unsupported_version", "Protocol version is not supported.");
  }
  if (record.success === true) {
    return parseDeviceSuccessResponse(record, options) as DeviceSuccessResponse<TResult>;
  }
  if (record.success === false) {
    return parseDeviceFailureResponse(record, options);
  }
  throw new ProtocolError("invalid_response", "Device response success flag is invalid.");
}

export const DEVICE_RESPONSE_ENVELOPE_FIELDS = {
  success: ["id", "version", "success", "method", "result"],
  failure: ["id", "version", "success", "method", "error"],
  error: ["code", "message", "retryable"],
} as const;

export interface CoreContractManifest {
  readonly protocolVersion: typeof PROTOCOL_VERSION;
  readonly methods: readonly DeviceMethodRow[];
  readonly responseEnvelope: typeof DEVICE_RESPONSE_ENVELOPE_FIELDS;
  readonly errors: readonly DeviceErrorRow[];
}

export function createCoreContractManifest(): CoreContractManifest {
  return {
    protocolVersion: PROTOCOL_VERSION,
    methods: DEVICE_METHOD_ROWS,
    responseEnvelope: DEVICE_RESPONSE_ENVELOPE_FIELDS,
    errors: DEVICE_ERROR_ROWS,
  };
}

function validateSessionRule(rule: DeviceSessionRule, hasSessionId: boolean, sessionId: unknown): void {
  if (rule === "forbidden") {
    if (hasSessionId) {
      throw new ProtocolError("invalid_request", "Device request must not include sessionId.");
    }
    return;
  }
  if (!hasSessionId) {
    if (rule === "required") {
      throw new ProtocolError("invalid_session", "Device request session is required.");
    }
    return;
  }
  if (!isSessionId(sessionId)) {
    throw new ProtocolError("invalid_session", "Device request session is invalid.");
  }
}

function validatePayloadRule(rule: DevicePayloadRule, hasPayload: boolean): void {
  if (rule === "forbidden" && hasPayload) {
    throw new ProtocolError("invalid_request", "Device request must not include payload.");
  }
  if (rule === "required" && !hasPayload) {
    throw new ProtocolError("invalid_params", "Device request payload is required.");
  }
}

function parseDeviceSuccessResponse(
  record: Record<string, unknown>,
  options: ParseDeviceResponseOptions,
): DeviceSuccessResponse {
  if (!hasOnlyObjectKeys(record, ["id", "version", "success", "method", "result"])) {
    throw new ProtocolError("invalid_response", "Device success response shape is invalid.");
  }
  if (!isDeviceMethod(record.method)) {
    throw new ProtocolError("invalid_response", "Device success response method is invalid.");
  }
  if (options.expectedMethod !== undefined && record.method !== options.expectedMethod) {
    throw new ProtocolError("invalid_response", "Device response method does not match request.");
  }
  if (!Object.prototype.hasOwnProperty.call(record, "result")) {
    throw new ProtocolError("invalid_response", "Device success response result is required.");
  }
  const parser = options.resultParsers?.[record.method];
  const result = parser === undefined ? record.result : parser(record.result);
  return {
    ...(Object.prototype.hasOwnProperty.call(record, "id") ? { id: record.id as string } : {}),
    version: PROTOCOL_VERSION,
    success: true,
    method: record.method,
    result,
  };
}

function parseDeviceFailureResponse(
  record: Record<string, unknown>,
  options: ParseDeviceResponseOptions,
): DeviceFailureResponse {
  if (!hasOnlyObjectKeys(record, ["id", "version", "success", "method", "error"])) {
    throw new ProtocolError("invalid_response", "Device failure response shape is invalid.");
  }
  if (Object.prototype.hasOwnProperty.call(record, "method")) {
    if (!isDeviceMethod(record.method)) {
      throw new ProtocolError("invalid_response", "Device failure response method is invalid.");
    }
    if (options.expectedMethod !== undefined && record.method !== options.expectedMethod) {
      throw new ProtocolError("invalid_response", "Device response method does not match request.");
    }
  }
  const error = parseDeviceErrorObject(record.error);
  return {
    ...(Object.prototype.hasOwnProperty.call(record, "id") ? { id: record.id as string } : {}),
    version: PROTOCOL_VERSION,
    success: false,
    ...(Object.prototype.hasOwnProperty.call(record, "method") ? { method: record.method as DeviceMethod } : {}),
    error,
  };
}

function parseDeviceErrorObject(value: unknown): DeviceError {
  const record = asContractRecord(value, "DeviceResponse.error", "invalid_response");
  if (!hasOnlyObjectKeys(record, ["code", "message", "retryable"])) {
    throw new ProtocolError("invalid_response", "Device response error shape is invalid.");
  }
  if (!isDeviceErrorCode(record.code)) {
    throw new ProtocolError("invalid_response", "Device response error code is invalid.");
  }
  const row = deviceErrorRow(record.code);
  if (record.message !== row.message || record.retryable !== row.retryable) {
    throw new ProtocolError("invalid_response", "Device response error does not match the contract table.");
  }
  return { code: row.code, message: row.message, retryable: row.retryable };
}

function asContractRecord(value: unknown, label: string, code: string): Record<string, unknown> {
  if (!isRecord(value)) {
    throw new ProtocolError(code, `${label} must be an object.`);
  }
  return value;
}

File: firmware/src/common/protocol/device_contract.cpp
#include "device_contract.h"

#include <string.h>

namespace signing {
namespace {

constexpr DeviceMethodRow kDeviceMethods[] = {
    {"get_status", DeviceSessionRule::forbidden, DevicePayloadRule::forbidden, "no payload", "StatusResult", "status_material_consistency"},
    {"identify_device", DeviceSessionRule::forbidden, DevicePayloadRule::required, "IdentifyDevicePayload", "IdentifyDeviceResult", "identification_display"},
    {"connect", DeviceSessionRule::forbidden, DevicePayloadRule::required, "ConnectPayload", "ConnectResult", "session_recovery_or_material_bootstrap_approval"},
    {"disconnect", DeviceSessionRule::required, DevicePayloadRule::forbidden, "no payload", "DisconnectResult", "active_session_cleanup"},
    {"get_capabilities", DeviceSessionRule::required, DevicePayloadRule::forbidden, "no payload", "CapabilitiesResult", "active_session_target_capability_state"},
    {"get_accounts", DeviceSessionRule::required, DevicePayloadRule::forbidden, "no payload", "AccountsResult", "active_session_account_availability"},
    {"policy_get", DeviceSessionRule::required, DevicePayloadRule::forbidden, "no payload", "PolicyGetResult", "active_session_policy_store"},
    {"get_approval_history", DeviceSessionRule::required, DevicePayloadRule::required, "ApprovalHistoryPayload", "ApprovalHistoryResult", "active_session_approval_history_store"},
    {"sign_transaction", DeviceSessionRule::required, DevicePayloadRule::required, "SuiSignTransactionPayload", "SignTransactionResult", "active_session_account_binding_signing_authorization"},
    {"sign_personal_message", DeviceSessionRule::required, DevicePayloadRule::required, "SuiSignPersonalMessagePayload", "SignPersonalMessageResult", "active_session_account_binding_user_confirmation"},
    {"policy_propose", DeviceSessionRule::required, DevicePayloadRule::required, "PolicyProposePayload", "PolicyProposalOutcome", "active_session_policy_review_auth_commit"},
    {"credential_prepare", DeviceSessionRule::required, DevicePayloadRule::required, "CredentialPreparePayload", "CredentialPreparation", "active_session_credential_preparation"},
    {"credential_propose", DeviceSessionRule::required, DevicePayloadRule::required, "CredentialProposePayload", "CredentialProposalOutcome", "active_session_credential_review_auth_commit"},
    {"get_result", DeviceSessionRule::required, DevicePayloadRule::required, "RetainedResponsePayload", "retained method response", "active_session_retained_response_lookup"},
    {"ack_result", DeviceSessionRule::required, DevicePayloadRule::required, "RetainedResponsePayload", "AckResult", "active_session_retained_response_cleanup"},
};

constexpr DeviceErrorRow kDeviceErrors[] = {
    {"invalid_request", false, "Device request is malformed.", "Envelope, JSON, id, version, or method shape is invalid."},
    {"invalid_params", false, "Device request payload is invalid.", "Method payload shape or value is invalid."},
    {"invalid_state", false, "Device state does not allow this request.", "Firmware state does not allow the method."},
    {"invalid_session", false, "Session is missing, expired, or does not match.", "Session is missing, expired, or does not match."},
    {"request_id_conflict", false, "Request id is already bound to a different request.", "Request id is already bound to a different retained request identity."},
    {"unknown_request", false, "Requested retained response does not exist.", "Requested retained response does not exist in the current session."},
    {"no_active_device", false, "No active device is configured.", "Host process has no selected active device for the requested scope."},
    {"device_not_found", true, "Requested device is not known to Agent-Q.", "Host process cannot find the requested stored device."},
    {"invalid_device_id", false, "Device id is invalid.", "Host-side device id input is invalid."},
    {"invalid_device", false, "Device identity is unsafe.", "Host process rejected a device identity as unsafe."},
    {"device_mismatch", false, "Connected device does not match the requested device.", "Host process connected to a device whose Firmware identity did not match the requested device id."},
    {"unsupported_method", false, "Device method is not supported.", "Method is not implemented by this Firmware or host boundary."},
    {"unsupported_version", false, "Protocol version is not supported.", "Protocol version is not supported."},
    {"unsupported_chain", false, "Chain is not supported.", "Chain is not supported for the method."},
    {"unsupported_transaction", false, "Transaction shape is not supported.", "Transaction shape is not supported by the current signing path."},
    {"malformed_transaction", false, "Transaction bytes are malformed.", "Transaction bytes or transaction structure are malformed."},
    {"payload_too_large", false, "Payload exceeds the current device payload capacity.", "Payload exceeds the current device payload capacity."},
    {"payload_unavailable", false, "Referenced payload is unavailable.", "Referenced internal payload is missing, expired, consumed, or wrong-session."},
    {"payload_conflict", true, "Another sensitive request blocks this payload.", "Another payload or sensitive flow blocks this request."},
    {"busy", true, "Device is busy with another request.", "Device is already handling another request."},
    {"timeout", true, "The signing request timed out on the device.", "Device or user action timed out."},
    {"user_rejected", false, "The signing request was rejected on the device.", "Device-local user confirmation rejected the request."},
    {"policy_rejected", false, "The signing request was rejected by device policy.", "Firmware policy rejected the request."},
    {"signing_failed", false, "The device could not produce a signature.", "Firmware could not produce the requested signature."},
    {"ui_error", false, "Device could not display or manage required UI.", "Firmware could not display or manage required device UI."},
    {"auth_unavailable", false, "Device-local authentication verifier is unavailable.", "Device-local authentication verifier is unavailable."},
    {"account_unavailable", false, "Device account material is unavailable or does not match the request.", "Device account material is unavailable or does not match the request."},
    {"policy_unavailable", false, "Firmware policy store or active policy is unavailable.", "Firmware policy store or active policy is unavailable."},
    {"history_unavailable", false, "Approval history is unavailable.", "Approval history is unavailable."},
    {"rng_unavailable", true, "Device random generator is unavailable.", "Firmware random generator is unavailable."},
    {"invalid_response", true, "Device response is malformed.", "Firmware response JSON, envelope, or method result shape is invalid."},
    {"handshake_failed", true, "Device status or identification handshake failed.", "Device status or identification handshake failed."},
    {"port_not_found", true, "Requested device port is not connected.", "Requested device port is not connected."},
    {"port_in_use", true, "Requested device port is already in use.", "Requested device port is already in use."},
    {"port_permission_denied", false, "Host process lacks permission to access the device port.", "Host process lacks permission to access the device port."},
    {"unsupported_transport", false, "Host environment does not support the required device transport.", "Host environment does not provide the transport required by the adapter."},
    {"transport_closed", true, "Device connection closed before a valid response.", "Established host-device connection closed before a valid response."},
    {"transport_error", true, "Host-device transport failed before a valid response.", "Host-device transport failed before a valid device response was received."},
    {"local_server_unavailable", true, "Local Agent-Q server is unavailable.", "Local CLI client cannot reach the local Agent-Q server."},
    {"identification_code_exhausted", true, "Identification code could not be allocated.", "Host process could not allocate a unique identification code."},
    {"internal_output_error", false, "Agent-Q produced an unexpected internal result.", "Output sanitization or internal result validation failed."},
    {"unknown_error", false, "Agent-Q request failed.", "Failure did not match a known public code."},
};

constexpr const char* kDeviceResponseSuccessFields[] = {
    "id", "version", "success", "method", "result",
};

constexpr const char* kDeviceResponseFailureFields[] = {
    "id", "version", "success", "method", "error",
};

constexpr const char* kDeviceResponseErrorFields[] = {
    "code", "message", "retryable",
};

}  // namespace

const DeviceMethodRow* device_method_row(const char* method)
{
    if (method == nullptr) {
        return nullptr;
    }
    for (const DeviceMethodRow& row : kDeviceMethods) {
        if (strcmp(row.method, method) == 0) {
            return &row;
        }
    }
    return nullptr;
}

const DeviceErrorRow* device_error_row(const char* code)
{
    if (code == nullptr) {
        return nullptr;
    }
    for (const DeviceErrorRow& row : kDeviceErrors) {
        if (strcmp(row.code, code) == 0) {
            return &row;
        }
    }
    return nullptr;
}

const DeviceMethodRow* device_method_rows(size_t* count)
{
    if (count != nullptr) {
        *count = sizeof(kDeviceMethods) / sizeof(kDeviceMethods[0]);
    }
    return kDeviceMethods;
}

const DeviceErrorRow* device_error_rows(size_t* count)
{
    if (count != nullptr) {
        *count = sizeof(kDeviceErrors) / sizeof(kDeviceErrors[0]);
    }
    return kDeviceErrors;
}

const char* const* device_response_success_fields(size_t* count)
{
    if (count != nullptr) {
        *count = sizeof(kDeviceResponseSuccessFields) / sizeof(kDeviceResponseSuccessFields[0]);
    }
    return kDeviceResponseSuccessFields;
}

const char* const* device_response_failure_fields(size_t* count)
{
    if (count != nullptr) {
        *count = sizeof(kDeviceResponseFailureFields) / sizeof(kDeviceResponseFailureFields[0]);
    }
    return kDeviceResponseFailureFields;
}

const char* const* device_response_error_fields(size_t* count)
{
    if (count != nullptr) {
        *count = sizeof(kDeviceResponseErrorFields) / sizeof(kDeviceResponseErrorFields[0]);
    }
    return kDeviceResponseErrorFields;
}

const char* device_session_rule_name(DeviceSessionRule rule)
{
    switch (rule) {
        case DeviceSessionRule::forbidden:
            return "forbidden";
        case DeviceSessionRule::required:
            return "required";
        case DeviceSessionRule::optional:
            return "optional";
    }
    return "";
}

const char* device_payload_rule_name(DevicePayloadRule rule)
{
    switch (rule) {
        case DevicePayloadRule::forbidden:
            return "forbidden";
        case DevicePayloadRule::required:
            return "required";
        case DevicePayloadRule::optional:
            return "optional";
    }
    return "";
}

}  // namespace signing
